cob_scan_unifier/src/scan_unifier_node.cpp

```diff
@@ -123,150 +166,99 @@ void scan_unifier_node::getParams()
- */
-void scan_unifier_node::topicCallbackLaserScan(const sensor_msgs::LaserScan::ConstPtr& scan_in, int scan_id)
+void ScanUnifierNode::messageFilterCallback(const sensor_msgs::LaserScan::ConstPtr& first_scanner, const sensor_msgs::LaserScan::ConstPtr& second_scanner, const sensor_msgs::LaserScan::ConstPtr& third_scanner)
 {
-  ROS_DEBUG_STREAM("scan-unifier: Laser-callback!");
-  if(!get_new_msg_received(scan_id))
-  {
-    vec_laser_struct_.at(scan_id).current_scan_msg = *scan_in;
-    ROS_DEBUG_STREAM("received msg from scan_id: " << scan_id);
-    set_new_msg_received(true, scan_id);
-  }
-  else
-  {
-    ROS_DEBUG_STREAM("received msg from scan_id: " << scan_id << "but storage blocked by mutex");
-  }
+  std::vector<sensor_msgs::LaserScan::ConstPtr> current_scans;
+  current_scans.push_back(first_scanner);
+  current_scans.push_back(second_scanner);
+  current_scans.push_back(third_scanner);
+
+  sensor_msgs::LaserScan unified_scan = unifyLaserScans(current_scans);
+
+  ROS_DEBUG("Publishing unified scan.");
+  topicPub_LaserUnified_.publish(unified_scan);
 }
 
 /**
- * @function unifieLaserScans
+ * @function unifyLaserScans
  * @brief unifie the scan information from all laser scans in vec_laser_struct_
  *
  * input: -
  * output:
  * @param: a laser scan message containing unified information from all scanners
  */
-sensor_msgs::LaserScan scan_unifier_node::unifieLaserScans()
+sensor_msgs::LaserScan ScanUnifierNode::unifyLaserScans(std::vector<sensor_msgs::LaserScan::ConstPtr> current_scans)
 {
   sensor_msgs::LaserScan unified_scan = sensor_msgs::LaserScan();
   std::vector<sensor_msgs::PointCloud> vec_cloud;
   vec_cloud.assign(config_.number_input_scans, sensor_msgs::PointCloud());
 
-  if(!vec_laser_struct_.empty())
+  if(!current_scans.empty())
   {
     ROS_DEBUG("start converting");
     for(int i=0; i < config_.number_input_scans; i++)
     {
-      vec_cloud.at(i).header.stamp = vec_laser_struct_.at(i).current_scan_msg.header.stamp;
-      ROS_DEBUG_STREAM("Converting scans to point clouds at index: " << i << ", at time: " << vec_laser_struct_.at(i).current_scan_msg.header.stamp << " now: " << ros::Time::now());
+      vec_cloud.at(i).header.stamp = current_scans.at(i)->header.stamp;
+      ROS_DEBUG_STREAM("Converting scans to point clouds at index: " << i << ", at time: " << current_scans.at(i)->header.stamp << " now: " << ros::Time::now());
       try
       {
-        listener_.waitForTransform("/base_link", vec_laser_struct_.at(i).current_scan_msg.header.frame_id,
-            vec_laser_struct_.at(i).current_scan_msg.header.stamp, ros::Duration(3.0));
+        if (!listener_.waitForTransform(frame_, current_scans.at(i)->header.frame_id,
+                                        current_scans.at(i)->header.stamp, ros::Duration(3.0)))
+        {
+          continue;
+        }
+        else
+        {
+          ROS_WARN_STREAM("Scan unifier skipped scan with " << current_scans.at(i)->header.stamp << " stamp.");
+        }
```

Comment:
Well, now that I'm looking at it:
**We** are handing over the `tf::TransformListener` object to the `projector_`. So in case all the `waitForTransform` magic we do in order to make the `transformLaserScanToPointCloud` work....maybe we just need to make our `listener_` object work!

Because currently it's using the default constructor (http://docs.ros.org/indigo/api/tf/html/c++/classtf_1_1TransformListener.html), i.e. default values for `max_cache_time`...maybe we can tweak something here...what also helps is "wait a little" in the constructor to let the listener_ fill its cache...as we are seeing this problem only in the very beginning...